File: sdks/go/webhook.go
package palai

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"strings"
	"time"
)

// The receiver-side webhook helper (spec §21.5/§23.10): a Go server that receives Palai's
// tool-call callbacks verifies the HMAC-SHA-256 signature over the EXACT raw body, enforcing the
// timestamp replay window and accepting a rotation-overlap header that carries several v1= values.
// This is the SAME routine the server's outbound signer computes, mirrored SDK-side and stdlib-only
// so a caller adds no dependency. It is a server-side helper by design; there is no browser path.
//
// The TS SDK ships no webhook verify (browser-relay stance), so the shared conformance corpus's
// signature-verify category had only the reference implementation until this Go leg — the Go SDK is
// the first shipped SDK to give that category a second independent impl (plan §T4 "signature verify").

const webhookSignatureVersion = "v1"

// SignWebhook computes the hex HMAC-SHA-256 over the signed input (version, delivery id, timestamp,
// then the exact raw body, joined by "."), matching the server's signer. Binding the version and id
// defeats cross-context replay.
func SignWebhook(secret []byte, deliveryID string, ts time.Time, rawBody []byte) string {
	mac := hmac.New(sha256.New, secret)
	mac.Write([]byte(webhookSignatureVersion + "." + deliveryID + "." + strconv.FormatInt(ts.Unix(), 10) + "."))
	mac.Write(rawBody)
	return hex.EncodeToString(mac.Sum(nil))
}

// VerifyWebhook is the receiver-side check: it recomputes the MAC over the raw body under secret,
// compares in constant time, and enforces the timestamp tolerance around now (the replay window).
// It accepts a header carrying several space-separated v1= values, so a receiver still on the old
// secret verifies a rotation-overlap attempt.
func VerifyWebhook(secret []byte, deliveryID string, ts time.Time, rawBody []byte, header string, now time.Time, tolerance time.Duration) bool {
	if skew := now.Sub(ts); skew > tolerance || skew < -tolerance {
		return false // outside the replay window
	}
	want := []byte(SignWebhook(secret, deliveryID, ts, rawBody))
	for _, field := range strings.Fields(header) {
		value, ok := strings.CutPrefix(field, webhookSignatureVersion+"=")
		if !ok {
			continue
		}
		if hmac.Equal([]byte(value), want) {
			return true
		}
	}
	return false
}

File: sdks/python/src/palai/webhook.py
"""Inbound webhook signature verification (§21.5, §23.10) — the receiver-side helper the TS SDK
does not ship, so the Python leg covers the ``signature-verify`` corpus category the TS runner omits.

It mirrors the server's ``adapters/integrations/webhook`` signer byte-for-byte: the signed input is
``"v1." + delivery_id + "." + unix + "."`` followed by the EXACT raw body, HMAC-SHA-256, hex-encoded.
A rotation header carries several space-separated ``v1=`` values so a receiver on either the old or
new secret verifies during the overlap.
"""

from __future__ import annotations

import hashlib
import hmac

_VERSION = "v1"


def sign(secret: bytes, delivery_id: str, timestamp: int, raw_body: bytes) -> str:
    """Compute the hex HMAC-SHA-256 over the signed input (version, delivery id, timestamp, body)."""
    prefix = f"{_VERSION}.{delivery_id}.{timestamp}.".encode()
    return hmac.new(secret, prefix + raw_body, hashlib.sha256).hexdigest()


def verify(
    secret: bytes,
    delivery_id: str,
    timestamp: int,
    raw_body: bytes,
    header: str,
    now: int,
    tolerance_seconds: int,
) -> bool:
    """Recompute the MAC under ``secret`` and compare in constant time, enforcing the configurable
    timestamp tolerance around ``now`` (the replay window). Accepts a header with several ``v1=``
    values (rotation overlap). ``timestamp``/``now`` are unix seconds; ``tolerance_seconds`` bounds
    ``|now - timestamp|``."""
    skew = now - timestamp
    if skew > tolerance_seconds or skew < -tolerance_seconds:
        return False  # outside the replay window
    want = sign(secret, delivery_id, timestamp, raw_body)
    for field in header.split():
        if not field.startswith(_VERSION + "="):
            continue
        value = field[len(_VERSION) + 1 :]
        if hmac.compare_digest(value, want):
            return True
    return False
